src/debugAdapter/goDebug.ts

```diff
@@ -479,71 +487,93 @@ class Delve {
+					return err ? reject(err) : resolve(res);
 				});
 			}, err => {
 				reject(err);
 			});
 		});
 	}
 
-	close(): Thenable<void> {
+	/**
+	 * To close the debugging session we do the following; If its a local process,
+	 * we try to halt and detach within a timeout otherwise we kill the debugger. If its a remote
+	 * process we try to halt and restart the remote target process running under the debugger.
+	 */
+	public close(): Thenable<void> {
 		if (this.noDebug) {
 			// delve isn't running so no need to halt
 			return Promise.resolve();
 		}
 		log('HaltRequest');
 
-		return new Promise(resolve => {
-			let timeoutToken: NodeJS.Timer;
-			if (this.debugProcess) {
-				timeoutToken = setTimeout(() => {
-					log('Killing debug process manually as we could not halt and detach delve in time');
-					killTree(this.debugProcess.pid);
-					resolve();
-				}, 1000);
-			}
+		return new Promise(async resolve => {
+			const timeoutToken: NodeJS.Timer = !this.isRemoteDebugging() && setTimeout(() => {
+				log('Killing debug process manually as we could not halt delve in time');
+				killTree(this.debugProcess.pid);
+				this.ensureDebugeeExecutableIsRemoved();
+				resolve();
+			}, 1000);
 
-			this.callPromise('Command', [{ name: 'halt' }]).then(() => {
-				if (timeoutToken) {
-					clearTimeout(timeoutToken);
-				}
+			let errMsg;
+			try {
+				await this.callPromise('Command', [{ name: 'halt' }]);
+			} catch (err) {
 				log('HaltResponse');
-				if (!this.debugProcess) {
-					log('RestartRequest');
-					return this.callPromise('Restart', this.isApiV1 ? [] : [{ position: '', resetArgs: false, newArgs: [] }])
-						.then(null, err => {
-							log('RestartResponse');
-							logError(`Failed to restart - ${(err || '').toString()}`);
-						})
-						.then(() => resolve());
-				} else {
-					log('DetachRequest');
-					return this.callPromise('Detach', [this.isApiV1 ? true : { Kill: true }])
-						.then(null, err => {
-							log('DetachResponse');
-							logError(`Killing debug process manually as we failed to detach - ${(err || '').toString()}`);
-							killTree(this.debugProcess.pid);
-						})
-						.then(() => resolve());
-				}
-			}, err => {
-				const errMsg = err ? err.toString() : '';
-				log('Failed to halt - ' + errMsg.toString());
-				if (errMsg.endsWith('has exited with status 0')) {
-					if (timeoutToken) {
-						clearTimeout(timeoutToken);
-					}
-					return resolve();
+				errMsg = err ? err.toString() : '';
+				log(`Failed to halt - ${errMsg}`);
+			}
+			clearTimeout(timeoutToken);
+
+			if (this.shouldDetach(errMsg)) {
+				log('DetachRequest');
+				try {
+					await this.callPromise('Detach', [this.isApiV1 ? true : { Kill: true }]);
+				} catch (err) {
+					log('DetachResponse');
+					logError(`Failed to detach - ${(err.toString() || '')}`);
 				}
-			});
+			}
+			if (!this.isRemoteDebugging()) {
+				await this.ensureDebugeeExecutableIsRemoved();
```

Comment:
If detach fails I wonder if we can actually delete the debugee.  Presumably the debugger has an open handle to the debugee so it should fail.  Seems like we shouldn't clear the timer until we know that halt and detach have succeeded.

Reply:
The timer should be cleared right after Halt, because it is to guard against the case where Halt request doesnt return in time.

But you bring up a good point. 

In case for local debugging, if detach fails, we should run the `killTree` to kill the process before deleting the file

Reply:
For local debugging I think we can simplify the entire flow.  Note that this assumes that the end result is that dlv and the debugee have been terminated and that the debug file has been deleted.
1. start a timer that will kill the process tree and delete the debugee; this MUST be no-fail
2. call halt and once that returns call detach; if either of these fail do nothing as the timer will clean up
3. do NOT clear the timer, if #2 succeeded it's a no-op
Is there any reason why this approach won't work?

Reply:
@ramya-rao-a  thoughts ?

Reply:
About step 2 & 3:

Halt is known to not return immediately based on the program being debugged. That was the reason I added the timer. When halt succeeds (or target exited) and we run `detach` i.e happy case, I want to allow detach the time it needs to successfully detach. We dont want the timer to timeout and do forceful kill/remove when detach is running.

If halt fails (but not because target exited) or if detach fails, then we want the forceful kill/remove.

If detach succeeds but doesnt remove the debuggee file, I don't want us to forcefully remove the debuggee file. If we see this, we should log a bug in delve.

Reply:
updated the logic, i think it should follow the conditions specified:
1. cleartimeout after halt
2. if localDebug && (halt fails because of anything else than target has exited || detach fails) => force kill  + force remove

Reply:
> We dont want the timer to timeout and do forceful kill/remove when detach is running.
Why?  The end result is the same either way and while I agree that we should do best effort to kill things gracefully I'm not convinced that the extra complexity to cover every corner case is worth it.

Reply:
@jhendrixMSFT Because I'd much rather have delve to the clean up (in the detach request), than kill it when the detach request is running.
If for what ever reason, the detach request doesnt complete in a reasonable time, then that would be a bug with delve and we will have a chance to report it.